src/System.Management.Automation/engine/Modules/GetModuleCommand.cs

```diff
@@ -545,22 +546,36 @@ private void GetLoadedModules(string[] names, IDictionary<string, ModuleSpecific
 
             foreach (PSModuleInfo module in modules)
             {
-                // TODO:
-                // moduleSpecification.Name may be a path and will not match module.Name when they refer to the same module.
-                // This actually causes the module to be returned always, so other specification checks are skipped erroneously.
-                // Instead we need to be able to look up or match modules by path as well (e.g. a new comparer for PSModuleInfo).
+                IEnumerable<ModuleSpecification> candidateModuleSpecs = GetCandidateModuleSpecs(moduleSpecificationTable, module);
 
-                // No table entry means we return the module
-                if (!moduleSpecificationTable.TryGetValue(module.Name, out ModuleSpecification moduleSpecification))
+                // Modules with table entries only get returned if they match them
+                // We skip the name check since modules have already been prefiltered base on the moduleSpec path/name
+                foreach (ModuleSpecification moduleSpec in candidateModuleSpecs)
                 {
-                    yield return module;
-                    continue;
+                    if (ModuleIntrinsics.IsModuleMatchingModuleSpec(module, moduleSpec, skipNameCheck: true))
+                    {
+                        yield return module;
+                    }
                 }
+            }
+        }
 
-                // Modules with table entries only get returned if they match them
-                if (ModuleIntrinsics.IsModuleMatchingModuleSpec(module, moduleSpecification))
+        /// <summary>
+        /// Take a dictionary of module specifications and return those that potentially match the module
+        /// passed in as a parameter (checks on names and paths).
+        /// </summary>
+        /// <param name="moduleSpecTable">The module specifications to filter candidates from.</param>
+        /// <param name="module">The module to find candidates for from the module specification table.</param>
+        /// <returns>The module specifications matching the module based on name, path and subpath.</returns>
+        private static IEnumerable<ModuleSpecification> GetCandidateModuleSpecs(
+            IDictionary<string, ModuleSpecification> moduleSpecTable,
+            PSModuleInfo module)
+        {
+            foreach (ModuleSpecification moduleSpec in moduleSpecTable.Values)
+            {
+                if (moduleSpec.Name == module.Name || moduleSpec.Name == module.Path || module.Path.Contains(moduleSpec.Name))
```

Comment:
FYI, this make existing cases of getting a module by name, using FullyQualifiedName case sensitive.  I'll submit a PR to fix this.